### PR TITLE
WIP: Add support for Lua 5.4

### PR DESCRIPTION
**Pull request by psychon**
No idea if the Travis build will work and I guess this should not be merged before Lua 5.4.0 is actually released.

### CODE REVIEW

**Comment by multiSnow** [comment]
[Lua 5.4 was released on 29 Jun 2020.](https://www.lua.org/versions.html#5.4)

File: .travis.yml [hunk 1]
```diff
@@ -36,6 +37,13 @@ install:
         && make SYSCFLAGS="-DLUA_USE_LINUX -ULUA_COMPAT_5_2 -DLUA_USE_APICHECK" SYSLIBS="-Wl,-E -ldl -lreadline" LUA_A=liblua.so MYCFLAGS="-fPIC" RANLIB=: AR="gcc -shared -ldl -o" liblua.so \
         && cd .. \
         && sudo make INSTALL_TOP=/usr/ INSTALL_INC=${LUAINCLUDE} TO_LIB=liblua.so linux install)
+    elif [[ "$LUA" == "5.4" ]]; then
+      travis_retry wget https://www.lua.org/work/lua-5.4.0-work2.tar.gz -O lua.tar.gz
```
Comment (gmbr3) [inline .travis.yml:41]:
```suggestion
      travis_retry wget https://www.lua.org/ftp/lua-5.4.0.tar.gz -O lua.tar.gz
```

Comment (gmbr3) [inline .travis.yml:41]:
```suggestion
      travis_retry wget https://www.lua.org/ftp/lua-5.4.0.tar.gz -O lua.tar.gz
```

File: .travis.yml [hunk 2]
```diff
@@ -36,6 +37,13 @@ install:
...
+      tar -xvzf lua.tar.gz
+      (cd lua-5.4.0-work2/src \
```
Comment (gmbr3) [inline .travis.yml:43]:
```suggestion
      (cd lua-5.4.0/src \
```

File: .travis.yml [hunk 3]
```diff
@@ -36,6 +37,13 @@ install:
         && make SYSCFLAGS="-DLUA_USE_LINUX -ULUA_COMPAT_5_2 -DLUA_USE_APICHECK" SYSLIBS="-Wl,-E -ldl -lreadline" LUA_A=liblua.so MYCFLAGS="-fPIC" RANLIB=: AR="gcc -shared -ldl -o" liblua.so \
```
Comment (gmbr3) [inline .travis.yml:37]:
```suggestion
        && make SYSCFLAGS="-DLUA_USE_LINUX -DLUA_COMPAT_5_2 -DLUA_USE_APICHECK" SYSLIBS="-Wl,-E -ldl -lreadline" LUA_A=liblua.so MYCFLAGS="-fPIC" RANLIB=: AR="gcc -shared -ldl -o" liblua.so \
```

File: .travis.yml [hunk 4]
```diff
@@ -36,6 +37,13 @@ install:
...
         && cd .. \
         && sudo make INSTALL_TOP=/usr/ INSTALL_INC=${LUAINCLUDE} TO_LIB=liblua.so linux install)
+    elif [[ "$LUA" == "5.4" ]]; then
+      travis_retry wget https://www.lua.org/work/lua-5.4.0-work2.tar.gz -O lua.tar.gz
+      tar -xvzf lua.tar.gz
+      (cd lua-5.4.0-work2/src \
+        && make SYSCFLAGS="-DLUA_USE_LINUX -ULUA_COMPAT_5_2 -DLUA_USE_APICHECK" SYSLIBS="-Wl,-E -ldl -lreadline" LUA_A=liblua.so MYCFLAGS="-fPIC" RANLIB=: AR="gcc -shared -ldl -o" liblua.so \
```
Comment (gmbr3) [inline .travis.yml:44]:
```suggestion
        && make SYSCFLAGS="-DLUA_USE_LINUX -DLUA_COMPAT_5_3 -DLUA_USE_APICHECK" SYSLIBS="-Wl,-E -ldl -lreadline" LUA_A=liblua.so MYCFLAGS="-fPIC" RANLIB=: AR="gcc -shared -ldl -o" liblua.so \
```

**Comment by psychon** [comment]
@gmbr3 Do you want to submit a new PR with your suggested changes? Waiting for me to update this PR would be waiting for a long time, sorry.

File: .travis.yml (hunk 4 above)
Comment (psychon) [reply .travis.yml:44]: Why? The idea is to catch uses of deprecated "stuff".

Comment (gmbr3) [reply .travis.yml:44]: -DLUA_COMPAT is always the previous version. So 5.4 it will be 5.3. It is in the lua source code.

**Comment by gmbr3** [comment]
OK, will do soon.

**Comment by gmbr3** [comment]
the test glib.lua hangs infinitely in lua 5.4 for some reason.

**Comment by gmbr3** [comment]
Leaving this up to everyone else

**Comment by psychon** [comment]
Thanks for giving it a try

**Comment by psychon** [comment]
`glib.lua` hands because Lua incorrectly throws a "C stack overflow" error.
<summary>backtrace at the time, showing that there is no overflow and Lua is mis-counting something
<details>

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7cce55b in __GI_abort () at abort.c:79
#2  0x0000555555561b7d in luaD_call ()
#3  0x000055555556e5d1 in luaV_execute ()
#4  0x00005555555617e7 in unroll ()
#5  0x0000555555560fa8 in luaD_rawrunprotected ()
#6  0x0000555555561e13 in lua_resume ()
#7  0x0000555555577b15 in auxresume ()
#8  0x0000555555577f15 in luaB_coresume ()
#9  0x0000555555561b7d in luaD_call ()
#10 0x000055555556e5d1 in luaV_execute ()
#11 0x0000555555561d17 in luaD_callnoyield ()
#12 0x0000555555560fa8 in luaD_rawrunprotected ()
#13 0x000055555556213e in luaD_pcall ()
#14 0x000055555555eae3 in lua_pcallk ()
#15 0x00007ffff7c9368d in closure_callback
    (cif=<optimized out>, ret=0x7fffffffdc20, args=0x7fffffffdaa0, closure_arg=0x7ffff6863550) at callable.c:1348
#16 0x00007ffff7ac0b02 in  () at /usr/lib/x86_64-linux-gnu/libffi.so.7
#17 0x00007ffff7ac0e60 in  () at /usr/lib/x86_64-linux-gnu/libffi.so.7
#18 0x00007ffff7b17054 in g_timeout_dispatch (source=0x5555555dbc80, callback=<optimized out>, user_data=<optimized out>)
    at ../../../glib/gmain.c:4800
#19 0x00007ffff7b164de in g_main_dispatch (context=0x55555564be80) at ../../../glib/gmain.c:3309
#20 g_main_context_dispatch (context=context@entry=0x55555564be80) at ../../../glib/gmain.c:3974
#21 0x00007ffff7b16890 in g_main_context_iterate
    (context=0x55555564be80, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>)
    at ../../../glib/gmain.c:4047
#22 0x00007ffff7b16b63 in g_main_loop_run (loop=0x555555720dc0) at ../../../glib/gmain.c:4241
#23 0x00007ffff7ac0ccd in  () at /usr/lib/x86_64-linux-gnu/libffi.so.7
#24 0x00007ffff7ac025a in  () at /usr/lib/x86_64-linux-gnu/libffi.so.7
#25 0x00007ffff7c94c41 in callable_call (L=<optimized out>) at callable.c:926
#26 0x0000555555561b7d in luaD_call ()
#27 0x000055555556e5d1 in luaV_execute ()
#28 0x000055555556e5d1 in luaV_execute ()
#29 0x000055555556e5d1 in luaV_execute ()
#30 0x000055555556e5d1 in luaV_execute ()
#31 0x0000555555561d17 in luaD_callnoyield ()
#32 0x0000555555560fa8 in luaD_rawrunprotected ()
#33 0x000055555556213e in luaD_pcall ()
#34 0x000055555555eae3 in lua_pcallk ()
#35 0x000055555555b86f in docall ()
#36 0x000055555555c59f in pmain ()
#37 0x0000555555561b7d in luaD_call ()
#38 0x0000555555561d17 in luaD_callnoyield ()
#39 0x0000555555560fa8 in luaD_rawrunprotected ()
#40 0x000055555556213e in luaD_pcall ()
#41 0x000055555555eae3 in lua_pcallk ()
#42 0x000055555555b65b in main ()
```

</details>
</summary>
<summary>I also tried writing a self-contained reproducer, but failed. Something is missing.
<details>

```c
#include <lua.h>
#include <lualib.h>
#include <lauxlib.h>
#include <stdlib.h>
#include <stdio.h>

const char lua_code[] =
"local coro;"
"coroutine.wrap(function()"
"  coro = coroutine.running();"
"  for i=1,1000 do"
"    start_new = function()"
"      coroutine.resume(coro)"
"    end;"
"    coroutine.yield(true)"
"  end;"
"  start_new = nil;"
"  return false;"
"end)()"
"main()"
;

static int main_cb(lua_State *L)
{
	int count = 0;
	for (;;) {
		int type = lua_getglobal(L, "start_new");
		if (type == LUA_TNIL)
			break;

		count++;
		printf("Iteration %d\n", count);

		lua_call(L, 0, 0);
	}
	return 0;
}

int main()
{
	int status;
	lua_State *L = luaL_newstate();
	luaL_openlibs(L);

	lua_pushglobaltable(L);
	lua_pushstring(L, "main");
	lua_pushcfunction(L, main_cb);
	lua_settable(L, -3);
	lua_settop(L, 0);

	status = luaL_dostring(L, lua_code);
	if (status) {
		printf("Status %d\n", status);
		printf("%s", lua_tostring(L, -1));
		lua_close(L);
		return 1;
	}
	lua_close(L);
	return 0;
}
```

</details>
</summary>

**Comment by psychon** [comment]
Superseded by #249 and the remaining issues are tracked in #247.

**Pull request closed without merging**